libdrgn/program.c

```diff
@@ -1108,6 +1108,135 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+{
+	if (arg->flags & SEARCH_ALL)
+		return true;
+	if (arg->flags & SEARCH_NAME)
+		if (name && arg->name && strcmp(name, arg->name) == 0)
+			return true;
+	if (arg->flags & SEARCH_ADDRESS)
+		if (arg->address >= addr && arg->address < addr + sym->st_size)
+			return true;
+	return false;
+}
+
+static int symbols_search_cb(Dwfl_Module *dwfl_module, void **userdatap,
+			     const char *module_name, Dwarf_Addr base,
+			     void *cb_arg)
+{
+	struct symbols_search_arg *arg = cb_arg;
+	int symtab_len, i;
+
+	symtab_len = dwfl_module_getsymtab(dwfl_module);
+	if (symtab_len == -1) {
+		arg->bad_symtabs = true;
+		return DWARF_CB_OK;
+	}
+	/* Ignore the zeroth null symbol */
+	for (i = 1; i < symtab_len; i++) {
+		GElf_Sym elf_sym;
+		GElf_Addr elf_addr;
+		const char *name;
+		struct drgn_symbol *sym;
+
+		name = dwfl_module_getsym_info(dwfl_module, i, &elf_sym,
+					       &elf_addr, NULL, NULL, NULL);
+		if (!symbol_match(arg, elf_addr, &elf_sym, name))
+			continue;
+
+		sym = malloc(sizeof(*sym));
+		if (!sym) {
+			arg->err = &drgn_enomem;
+			return DWARF_CB_ABORT;
+		}
+		drgn_symbol_from_elf(name, elf_addr, &elf_sym, sym);
+		if (!symbolp_vector_append(&arg->results, &sym)) {
+			drgn_symbol_destroy(sym);
+			arg->err = &drgn_enomem;
+			return DWARF_CB_ABORT;
+		}
+	}
+	return DWARF_CB_OK;
+}
+
+static struct drgn_error *
+symbols_search(struct drgn_program *prog, struct symbols_search_arg *arg,
+	       struct drgn_symbol ***syms_ret, size_t *count_ret)
+{
+	if (!prog->dbinfo)
+		return drgn_error_create(DRGN_ERROR_LOOKUP,
+					 "could not find matching symbols");
+
+	symbolp_vector_init(&arg->results);
+
+	if (dwfl_getmodules(prog->dbinfo->dwfl, symbols_search_cb,
+			    arg, 0)) {
+		int i;
+		for (i = 0; i < arg->results.size; i++) {
+			drgn_symbol_destroy(arg->results.data[i]);
+		}
+		symbolp_vector_deinit(&arg->results);
+		return arg->err;
+	}
+
+	if (arg->results.size == 0) {
+		return drgn_error_format(DRGN_ERROR_LOOKUP,
+					 "could not find matching symbols%s",
+					 arg->bad_symtabs ?
+					 " (could not get some symbol tables)" : "");
+	}
+	symbolp_vector_shrink_to_fit(&arg->results);
+	*count_ret = arg->results.size;
+	*syms_ret = arg->results.data;
+	return NULL;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_find_symbols_by_name(struct drgn_program *prog, const char *name,
+				  struct drgn_symbol ***syms_ret,
+				  size_t *count_ret)
+{
+	struct symbols_search_arg arg = {
+		.name = name,
+	};
+	if (name)
+		arg.flags = SEARCH_NAME;
+	else
+		arg.flags = SEARCH_ALL;
+	return symbols_search(prog, &arg, syms_ret, count_ret);
+}
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_find_symbols_by_address(struct drgn_program *prog,
+				     uint64_t address,
+				     struct drgn_symbol ***syms_ret,
+				     size_t *count_ret)
+{
+	struct symbols_search_arg arg = {
+		.address = address,
+		.flags = SEARCH_ADDRESS,
+	};
+	return symbols_search(prog, &arg, syms_ret, count_ret);
```

Comment:
Although the code reuse is nice, it seems a bit wasteful to check every module. I think I'd rather this only check the module found with `dwfl_addrmodule`.